doc/developer-notes.md

```diff
@@ -839,6 +840,36 @@ namespace {
 #endif // BITCOIN_FOO_BAR_H
 ```
 
+Construct On First Use
+------------------------
+
+It is not possible to control the order in which non-local static objects in *different* translation units are initialized.
+This is the so-called "static initialization order problem". It could cause the use of an object:
+- before it is constructed (on program startup)
+- after it is destructed (at program exit)
+
+Both cases lead to undefined behavior.
+
+To prevent the static initialization order problem, be sure to Construct On First Use:
```

Comment:
I don't think you should make this suggestion in general. This solution, while valuable in rare cases such as logging and tracking locks, is a brute-force one that does not belong in actual application code. It's *messy* and assumes we don't actually know what's going on or what are all the edge cases.

You're supposed to expliclty allocate things in a deterministic order during initialization and destroy it in a (usually reversed) deterministic order on shutdown, not rely on global initialization order nor first use.